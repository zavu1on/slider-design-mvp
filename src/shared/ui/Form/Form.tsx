import { type FormHTMLAttributes, type ReactNode, type Ref } from 'react';
import { type FieldValues, type UseFormReturn } from 'react-hook-form';
import { Form as UIForm } from '@/generated/shadcn/form';

type FormProps<Values extends FieldValues> =
  FormHTMLAttributes<HTMLFormElement> & {
    form: UseFormReturn<Values>;
    ref?: Ref<HTMLFormElement>;
  };

export const Form = <Values extends FieldValues = FieldValues>({
  form,
  children,
  ref,
  ...props
}: FormProps<Values>): ReactNode => {
  return (
    <UIForm {...form}>
      <form ref={ref} {...props}>
        {children}
      </form>
    </UIForm>
  );
};
