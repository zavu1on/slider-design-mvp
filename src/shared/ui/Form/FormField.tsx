'use client';

import { type ReactElement, type ReactNode, cloneElement } from 'react';
import { type Control, type FieldValues, type Path } from 'react-hook-form';
import {
  FormControl,
  FormDescription,
  FormItem,
  FormLabel,
  FormMessage,
  FormField as UIFormField,
} from '@/generated/shadcn/form';

export type FormChildrenProps<Values extends FieldValues> = {
  name: Path<Values>;
  label?: string;
  description?: string;
  control: Control<Values>;
  containerClassName?: string;

  // multi-select only
  onValueChange?: (value: string[]) => void;
  defaultValue?: string[];
};

type FormChildren<Values extends FieldValues> = ReactElement & {
  props: Omit<FormChildrenProps<Values>, 'control' | 'containerClassName'>;
};

type FormFieldProps<Values extends FieldValues> = {
  control: Control<Values>;
  children: FormChildren<Values>;
  className?: string;
};

export const FormField = <Values extends FieldValues = FieldValues>({
  control,
  children,
  className,
}: FormFieldProps<Values>): ReactNode => {
  if (!children.props?.name) throw Error('No "name" param!');

  return (
    <UIFormField
      control={control}
      name={children.props.name}
      render={({ field }) => (
        <FormItem className={className}>
          <FormLabel>{children.props.label}</FormLabel>
          <FormControl>
            {cloneElement(children, {
              onValueChange: field.onChange,
              defaultValue: field.value,
              ...children.props,
              ...field,
            })}
          </FormControl>
          <FormDescription>{children.props.description}</FormDescription>
          <FormMessage />
        </FormItem>
      )}
    />
  );
};
