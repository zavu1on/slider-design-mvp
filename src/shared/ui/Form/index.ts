export { Form } from './Form';
export { FormField, type FormChildrenProps } from './FormField';
