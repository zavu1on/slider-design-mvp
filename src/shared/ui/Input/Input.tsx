import { type InputHTMLAttributes, type ReactNode } from 'react';
import { type FieldValues } from 'react-hook-form';
import { Input as UIInput } from '@/generated/shadcn/input';
import { type FormChildrenProps, FormField } from '../Form';

type InputProps<Values extends FieldValues> = FormChildrenProps<Values> &
  InputHTMLAttributes<HTMLInputElement>;

export const Input = <Values extends FieldValues = FieldValues>({
  control,
  containerClassName,
  ...props
}: InputProps<Values>): ReactNode => {
  return (
    <FormField control={control} className={containerClassName}>
      <UIInput {...props} />
    </FormField>
  );
};
