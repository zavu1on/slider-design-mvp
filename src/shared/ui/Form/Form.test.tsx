import { yupResolver } from '@hookform/resolvers/yup';
import {
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useForm } from 'react-hook-form';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import yup from 'yup';
import { Input } from '../Input';
import { Form } from './Form';

const testFormSchema = yup.object({
  firstName: yup.string().required().min(3),
  lastName: yup.string().required().min(3),
});

type TestFormSchema = yup.InferType<typeof testFormSchema>;

describe('Form + Input', () => {
  const onSubmit = vi.fn();

  beforeEach(() => {
    cleanup();
  });

  const TestForm = () => {
    const form = useForm<TestFormSchema>({
      defaultValues: { firstName: '', lastName: '' },
      resolver: yupResolver(testFormSchema),
    });

    return (
      <Form form={form} onSubmit={form.handleSubmit(onSubmit)}>
        <Input control={form.control} name="firstName" label="First Name" />
        <Input control={form.control} name="lastName" label="Last Name" />
        <button type="submit">Submit</button>
      </Form>
    );
  };

  const TestFormWithError = () => {
    const form = useForm<TestFormSchema>({
      defaultValues: { firstName: '', lastName: '' },
      resolver: yupResolver(testFormSchema),
    });

    return (
      <Form form={form} onSubmit={form.handleSubmit(onSubmit)}>
        <Input control={form.control} name="firstName" label="First Name" />
        <Input control={form.control} label="Last Name" />
        <button type="submit">Submit</button>
      </Form>
    );
  };

  it('updates form values and calls onSubmit on submit', async () => {
    render(<TestForm />);

    const firstNameInput = screen.getByLabelText('First Name');
    const lastNameInput = screen.getByLabelText('Last Name');
    const submitButton = screen.getByRole('button', { name: /submit/i });

    fireEvent.change(firstNameInput, { target: { value: 'John' } });
    fireEvent.change(lastNameInput, { target: { value: 'Doe' } });

    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(onSubmit).toHaveBeenCalledOnce();
      expect(onSubmit).toHaveBeenCalledWith(
        { firstName: 'John', lastName: 'Doe' },
        expect.anything()
      );
    });
  });

  it('throws error if Input name is missing', () => {
    expect(() => render(<TestFormWithError />)).toThrowError();
  });

  it('updates form values on user input, calls onSubmit on Enter submit', async () => {
    render(<TestForm />);
    const firstNameInput = screen.getByLabelText('First Name');
    const lastNameInput = screen.getByLabelText('Last Name');

    await userEvent.type(firstNameInput, 'John');
    await userEvent.type(lastNameInput, 'Doe');

    expect(screen.getByDisplayValue('John')).toBeInTheDocument();
    expect(screen.getByDisplayValue('Doe')).toBeInTheDocument();

    await userEvent.keyboard('{enter}');

    expect(onSubmit).toHaveBeenCalledWith(
      { firstName: 'John', lastName: 'Doe' },
      expect.anything()
    );
  });

  it('show error message if form is invalid', async () => {
    const { container } = render(<TestForm />);
    const firstNameInput = screen.getByLabelText('First Name');
    const lastNameInput = screen.getByLabelText('Last Name');
    const submitButton = screen.getByRole('button', { name: /submit/i });

    await userEvent.type(firstNameInput, 'J');
    await userEvent.type(lastNameInput, 'D');
    await userEvent.click(submitButton);

    expect(
      screen.queryByText('firstName must be at least 3 characters')
    ).toBeInTheDocument();
    expect(
      screen.queryByText('lastName must be at least 3 characters')
    ).toBeInTheDocument();

    expect(container).toMatchSnapshot();

    await userEvent.clear(firstNameInput);
    await userEvent.clear(lastNameInput);

    await userEvent.type(firstNameInput, 'Joe');
    await userEvent.type(lastNameInput, 'D');
    await userEvent.click(submitButton);

    expect(
      screen.queryByText('firstName must be at least 3 characters')
    ).not.toBeInTheDocument();
    expect(
      screen.queryByText('lastName must be at least 3 characters')
    ).toBeInTheDocument();
  });

  it("doesn't call onSubmit if form is invalid", async () => {});
});
