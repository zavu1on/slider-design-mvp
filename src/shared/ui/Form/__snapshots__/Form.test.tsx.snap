// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Form + Input > show error message if form is invalid 1`] = `
<div>
  <form>
    <div
      class="grid gap-2"
      data-slot="form-item"
    >
      <label
        class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 data-[error=true]:text-destructive"
        data-error="true"
        data-slot="form-label"
        for="«r6»-form-item"
      >
        First Name
      </label>
      <input
        aria-describedby="«r6»-form-item-description «r6»-form-item-message"
        aria-invalid="true"
        class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
        data-slot="form-control"
        id="«r6»-form-item"
        label="First Name"
        name="firstName"
        value="J"
      />
      <p
        class="text-muted-foreground text-sm"
        data-slot="form-description"
        id="«r6»-form-item-description"
      />
      <p
        class="text-destructive text-sm"
        data-slot="form-message"
        id="«r6»-form-item-message"
      >
        firstName must be at least 3 characters
      </p>
    </div>
    <div
      class="grid gap-2"
      data-slot="form-item"
    >
      <label
        class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 data-[error=true]:text-destructive"
        data-error="true"
        data-slot="form-label"
        for="«r7»-form-item"
      >
        Last Name
      </label>
      <input
        aria-describedby="«r7»-form-item-description «r7»-form-item-message"
        aria-invalid="true"
        class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
        data-slot="form-control"
        id="«r7»-form-item"
        label="Last Name"
        name="lastName"
        value="D"
      />
      <p
        class="text-muted-foreground text-sm"
        data-slot="form-description"
        id="«r7»-form-item-description"
      />
      <p
        class="text-destructive text-sm"
        data-slot="form-message"
        id="«r7»-form-item-message"
      >
        lastName must be at least 3 characters
      </p>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
